### ISSUE TITLE
code issue

### ISSUE THREAD

**Issue by QingXiaoYan** [zdk258/CorrCLIP#6]
Can the code for Figure 2 (Impact of patch correlation scope on the performance of COCO Stuff) in the paper be made public?

<img width="860" height="752" alt="Image" src="https://github.com/user-attachments/assets/ada50c72-d525-4adb-857c-4293adadc414" />

**Comment by zdk258** [comment]
Sorry, this figure was generated with a standalone script from an earlier debugging stage, and it differs quite a lot from the current codebase. So it is a bit cumbersome for us to clean up and release that code directly. However, I can provide the pseudocode and the main steps for reproducing it.

For **Figure 2(a)**, to reproduce the **lowest** results, you can replace the original attention weights with the following:

```python
# x is DINO's features, H and W are the height and width of DINO's features
similarity = torch.einsum("m c, n c -> m n", x, x)
similarity_clone = similarity.clone()
similarity_ = similarity.clone()

# get the instance masks from labels.
# For intra-class, we use the labels of COCO-Stuff.
# For intra-object, we use the labels of COCO-Panoptic.
label_values = torch.unique(label)
region_masks = [(idx == label).view(-1) for idx in label_values]
equal_idx = torch.zeros_like(similarity).bool()
for mask in region_masks:
    equal_idx += torch.outer(mask, mask)

# take Lowest for an example
similarity_[equal_idx] = 100
k = int(H * W * ratio)
threshold_value = torch.kthvalue(similarity_, k, dim=-1, keepdim=True).values
sim_mask = similarity_ < threshold_value

similarity[~equal_idx] = float('-inf')
similarity[sim_mask] = similarity_clone[sim_mask]

attn_weights = similarity.softmax(dim=-1)
````

Then you can try different `ratio` values and evaluate the resulting mIoU on **COCO-Stuff**.
For **highest**, the procedure is similar, except that you keep the highest correlations instead.

For **intra-class**, you can replace `instance_mask` in the current code with the **COCO-Stuff** labels.
For **intra-object**, you can replace `instance_mask` with the **COCO-Panoptic** labels.

Please note that the above procedure does **not** use **feature refinement** or **map correction**.

For **Figure 2(b)**, you only need to compute, for each method, the sum of the final `attn_weights` corresponding to patch pairs that belong to different classes.

```
```

**Comment by QingXiaoYan** [comment]
Thank you for your reply.
